taar/recommenders/similarity_recommender.py

```diff
@@ -0,0 +1,153 @@
+
+S3_BUCKET = 'telemetry-private-analysis-2'
+DONOR_LIST_KEY = 'taar/legacy/addon_donors.json'
+LR_CURVES_SIMILARITY_TO_PROBABILITY = 'taar/legacy/test/lr_curves.json'
+
+CATEGORICAL_FEATURES = ["geoCity", "locale", "os"]
+CONTINUOUS_FEATURES = ["subsessionLength", "bookmark_count", "tab_open_count", "total_uri", "unique_tlds"]
+
+logger = logging.getLogger(__name__)
+
+
+class SimilarityRecommender(BaseRecommender):
+    """ A recommender class that returns top N addons based on the client similarity
+    with a set of candidate addon donors. Several telemetry fields are used to compute
+    pairwise similarity with the donors and similarities are converted into a likelihood
+    ratio of being a good match versus not being a good match. These quantities are then
+    used to rank specific addons for recommendation.
+
+    This will load a json file containing updated list of addon donors updated periodically
+    by a separate weekly process using Longitdudinal Telemetry data.
+
+    This recommender may provide useful recommendations when collaborative_recommender
+    may not work.
+    """
+
+    def __init__(self):
+        self.donors_list = []
+
+        # Download the addon donors list.
+        donors_pool = utils.get_s3_json_content(S3_BUCKET, DONOR_LIST_KEY)
+        if donors_pool is None:
+            logger.error("Cannot download the donor list: {}".format(DONOR_LIST_KEY))
+
+        for donor in donors_pool:
+            # Separate out the categorical and numerical features for the donor.
+            d = dict.fromkeys(['active_addons', 'categorical_features', 'continuous_features'])
+            d['categorical_features'] = [donor.get(specified_key) for specified_key in CATEGORICAL_FEATURES]
+            d['continuous_features'] = [donor.get(specified_key) for specified_key in CONTINUOUS_FEATURES]
+            d['active_addons'] = donor['activeAddons']
+            self.donors_list.append(d)
+
+        # Download the probability mapping curves from similarity to likelihood of being a good donor.
+        self.lr_curves = utils.get_s3_json_content(S3_BUCKET, LR_CURVES_SIMILARITY_TO_PROBABILITY)
+
+        if self.lr_curves is None:
+            logger.error("Cannot download the lr curves: {}".format(LR_CURVES_SIMILARITY_TO_PROBABILITY))
+
+    def can_recommend(self, client_data):
+        # We can't recommend if we don't have our data files.
+        if self.donors_list is None or self.lr_curves is None:
+            return False
+
+        # Check that the client info contains a non-None value for each required telemetry field.
+        REQUIRED_FIELDS = ["geoCity", "subsessionLength", "locale", "os", "bookmark_count", "tab_open_count",
+                           "total_uri", "unique_tlds"]
+
+        has_fields = all([client_data.get(f, None) is not None for f in REQUIRED_FIELDS])
+        if not has_fields:
+            logger.error("Unusable client data encountered in SimilarityRecommender, fields not complete")
```

Comment:
Let's report the client id here using `extra`. Also, no need to mention the name of the module: this is already automatically reported by the `logger`. I think this could simply be "Incomplete client data" or something like that.